File: code/temp.cpp
/**********************************************
  > File Name		: temp.cpp
  > Author		    : lunar
  > Email			: lunar_ubuntu@qq.com
  > Created Time	: Sun 02 Jan 2022 04:50:40 PM CST
  > Location        : Shanghai
  > Copyright@ https://github.com/xiaoqixian
 **********************************************/

/**
 * SSP: max flow based on Edmonds-Karp algorithm, shortest path based on Bellman-Ford.
 */
#include <vector>
#include <stdio.h>
#include <cstring>
#include <queue>
#include <unistd.h>

using namespace std;

#define N 250
#define M 250
#define INF 0x3f3f3f3f

struct qxx {
  int nex, t, v, c;
};
qxx e[M];
int h[N], cnt = 1;
void add_path(int f, int t, int v, int c) {
  e[++cnt] = (qxx){h[f], t, v, c}, h[f] = cnt;
}
void add_flow(int f, int t, int v, int c) {
  add_path(f, t, v, c);
  add_path(t, f, 0, -c);
}
int dis[N], pre[N], incf[N];
bool vis[N];
int s, t;
bool spfa() {
  memset(dis, 0x3f, sizeof(dis));
  queue<int> q;
  q.push(s), dis[s] = 0, incf[s] = INF, incf[t] = 0;
  while (q.size()) {
    int u = q.front();
    q.pop();
    vis[u] = 0;
    for (int i = h[u]; i; i = e[i].nex) {
      const int &v = e[i].t, &w = e[i].v, &c = e[i].c;
      if (!w || dis[v] <= dis[u] + c) continue;
      dis[v] = dis[u] + c, incf[v] = min(w, incf[u]), pre[v] = i;
      if (!vis[v]) q.push(v), vis[v] = 1;
    }
  }
  return incf[t];
}
int maxflow = 0, mincost = 0;
void update() {
  maxflow += incf[t];
  for (int u = t; u != s; u = e[pre[u] ^ 1].t) {
    e[pre[u]].v -= incf[t], e[pre[u] ^ 1].v += incf[t];
    mincost += incf[t] * e[pre[u]].c;
  }
}

int main() {
    add_path((int)'S', (int)'A', 8, 2);
    add_path((int)'S', (int)'B', 7, 8);
    add_path((int)'A', (int)'B', 5, 5);
    add_path((int)'A', (int)'C', 9, 2);
    add_path((int)'C', (int)'B', 2, 1);
    add_path((int)'B', (int)'D', 9, 3);
    add_path((int)'D', (int)'C', 6, 4);
    add_path((int)'D', (int)'T', 10, 7);
    add_path((int)'C', (int)'T', 5, 6);
    while (spfa()) {
        update();
    }
    printf("<%d, %d>\n", maxflow, mincost);
}

File: leetcode/temp.py
# !/usr/bin/python3
# -*- coding: utf-8 -*-
# > Author          : lunar
# > Email           : lunar_ubuntu@qq.com
# > Created Time    : Thu 31 Mar 2022 06:24:38 PM CST
# > Location        : Shanghai
# > Copyright@ https://github.com/xiaoqixian

commands = ["add","add","add","add","add","add","add","add","add","erase","search","add","erase","erase","erase","add","search","search","search","erase","search","add","add","add","erase","search","add","search","erase","search","search","erase","erase","add","erase","search","erase","erase","search","add","add","erase","erase","erase","add","erase","add","erase","erase","add","add","add","search","search","add","erase","search","add","add","search","add","search","erase","erase","search","search","erase","search","add","erase","search","erase","search","erase","erase","search","search","add","add","add","add","search","search","search","search","search","search","search","search","search"]

args = [[16],[5],[14],[13],[0],[3],[12],[9],[12],[3],[6],[7],[0],[1],[10],[5],[12],[7],[16],[7],[0],[9],[16],[3],[2],[17],[2],[17],[0],[9],[14],[1],[6],[1],[16],[9],[10],[9],[2],[3],[16],[15],[12],[7],[4],[3],[2],[1],[14],[13],[12],[3],[6],[17],[2],[3],[14],[11],[0],[13],[2],[1],[10],[17],[0],[5],[8],[9],[8],[11],[10],[11],[10],[9],[8],[15],[14],[1],[6],[17],[16],[13],[4],[5],[4],[17],[16],[7],[14],[1]]

for i in range(len(args)):
    print("sl.", commands[i], "(", args[i][0], ");", sep = "")
